### DISCUSSION TITLE
Canonical way to restart the application?

### DISCUSSION THREAD

**Discussion by tomk3003** [mojolicious/mojo#2045] [Q&A] [answered]
I am working on an application where the admin user needs to restart the application after changing the configuration through a configuration dialog. With hypnotoad it is easy to trigger a hot deployment with a signal to the process. 
But I have no idea how to do that for the other servers. The application admin will not have access to stop and start the app from the OS level.

TIA
Thomas

**Comment by Grinnz** [comment, marked as answer]
The other servers do not have process management, it is expected you will manage that as with any other service. If they will not have access to restart it as a system service then perhaps it should run as a user service.

**Reply by tomk3003** [reply]
Since the admins users are application admins only (without any access to the OS level), I'll just rely on hypnotoad and mock this for the tests.

But it would be nice to be able to have the server stop listening, create a new instance of the app object and start listening again with the new app. 

Thanks,
Thomas

**Reply by tomk3003** [reply]
I played around a bit and it looks that:
```perl
package MyApp;
use Mojo::Base 'Mojolicious', -signatures;
 
sub startup ($self) {
    $self->hook(
        before_server_start => sub ( $srv, $app ) {
            $app->{_server_instance} = $srv;
        }
    );
}
 
sub reload_app ($self) {
    return unless my $srv = $self->{_server_instance};
    $srv->stop;
    my $new = $srv->build_app(__PACKAGE__);
    $new->server($srv);
    $srv->start;
}
 ``` 
seems to do what I want, when I call "reload_app" from the controller. 
At least when using Mojo::Server::Deamon. 
But I am not sure whether this is safe and doesn't leak memory.

**Reply by kraih** [reply]
No, the other servers cannot restart themselves. You need something like systemd to do that for them.

**Reply by tomk3003** [reply]
After thinking about it, my approach seems to be not practical anyway. I should separate the configuration into the parts that should be editable by the app admins and the values that should be set by the OS admins. For changes to the former to take effect a reload of the page should most likely be enough.

**Reply by daleif** [reply]
> After thinking about it, my approach seems to be not practical anyway. I should separate the configuration into the parts that should be editable by the app admins and the values that should be set by the OS admins. For changes to the former to take effect a reload of the page should most likely be enough.

I got hit by something similar. Just make sure that nothing in `startup` depends on the values editable by the app admins as the `startup` often only runs once. As long as it is stuff that runs in a controller it's fine.
